Resolve #11727

wu-yafeng opened the pull request:
Summary of the changes (Less than 80 chars)
 - Add a BuildServiceProviderValidator for checking startup method 'ConfigureServices(IServiceCollection,..)'
 - Detail 2

Addresses #11727

Commit:
Resolve #11727

- Add a BuildServiceProviderValidator for checking startup method 'ConfigureServices(IServiceCollection,..)'

wu-yafeng commented:
It seems no RPs for this issue.So,I do it~ :smile:

Commit:
Merge branch 'master' into startup-analyze

pranavkm commented on src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs:
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
+            "Design",
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            helpLinkUri: "https://github.com/aspnet/AspNetCore/issues/11727"
```
@davidfowl suggestions on where this should point to?

davidfowl replied:
I would assume these should all point to a single docs article via an aka.ms link

rynowak replied:
This should be an aka.ms and it should point to a docs article.

rynowak replied:
@wu-yafeng - we'll get back to you with a URL to put here

rynowak replied:
@wu-yafeng - please make this point to `https://aka.ms/AA5k895`. I'll write a docs article as part of this release.

Commit:
Fix file name and class name not match

Commit:
Merge branch 'master' into startup-analyze

Tratcher commented:
Why are Startup analyzers in MVC?

rynowak commented:
There's a work item on my plate to get the more general `Microsoft.AspNetCore.Analzyers` assembly shipped in the SDK. It's a purely point in time issue.

rynowak commented on src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs:
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
+
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
+            "Cannot use BuildServiceProvider in ConfigureServices.",
```
Suggestion: `Do not call 'IServiceCollection.BuildServiceProvider' in 'ConfigureServices'.`

rynowak commented on src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs:
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
             helpLinkUri: "https://aka.ms/YJggeFn");
+
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
```
@Tratcher - raised an interesting point. Right now these build from inside the MVC solution - I'm planning to move these to a more general place, and I'll handle the update of this ID when I do it.

rynowak commented on src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs:
```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
+            "Cannot use BuildServiceProvider in ConfigureServices.",
+            "BuildServiceProvider called by runtime,don't call this method in your application code directly",
```
Suggestion: `Calling 'BuildServiceProvider' from application code results in an additional copy of singleton services being created. Consider alternatives such as dependency injecting services as parameters to `Configure`.

The docs article should cover:
- DI of parameters to configure
- Using DI with options https://andrewlock.net/access-services-inside-options-and-startup-using-configureoptions/
- @davidfowl  anything else?

davidfowl replied:
Does this warn if you return the IServiceProvider?

wu-yafeng replied:
No,this warn if called `BuildServiceProvider` in 'ConfigureServices'.

It seems add validators such as `ConfigureServiceMethodSignatureValidator` and `ConfigureMethodSignatureValidator` to check startup methods signatures

davidfowl replied:
This only affects 3.0 applications right?

rynowak replied:
These analyzers will only show up for 3.0+ because they ship in the SDK.

Commit:
Resolve conversation:
- helpLink point to 'aka.ms'
- Update DiagnosticDescriptor title
- Update DiagnosticDescriptor messageFormat

wu-yafeng commented:
There are no tests or asserts for `DiagnosticDescriptor` properties such as `Id`,`Title.
Should I add tests or asserts for them?

rynowak commented:
> Should I add tests or asserts for them?

I don't think it's necessary, those tests would just verify that the text is what you typed in. Those kinds of tests are valuable if the diagnostic output contains format parameters.

rynowak commented:
Thanks!
